llama_toolchain/inference/ollama.py

```diff
@@ -0,0 +1,264 @@
+        raise NotImplementedError()
+
+    def _messages_to_ollama_messages(self, messages: list[Message]) -> list:
+        ollama_messages = []
+        for message in messages:
+            ollama_messages.append(
+                {"role": message.role, "content": message.content}
+            )
+
+        return ollama_messages
+
+    async def chat_completion(self, request: ChatCompletionRequest) -> AsyncGenerator:
+        if not request.stream:
+            r = await self.client.chat(
+                model=self.model,
+                messages=self._messages_to_ollama_messages(request.messages),
+                stream=False,
+                #TODO: add support for options like temp, top_p, max_seq_length, etc
+            )
+            if r['done']:
+                if r['done_reason'] == 'stop':
+                    stop_reason = StopReason.end_of_turn
+                elif r['done_reason'] == 'length':
+                    stop_reason = StopReason.out_of_tokens
+
+            completion_message = decode_assistant_message_from_content(
+                r['message']['content'],
+                stop_reason,
+            )
+            yield ChatCompletionResponse(
+                completion_message=completion_message,
+                logprobs=None,
+            )
+        else:
+            yield ChatCompletionResponseStreamChunk(
+                event=ChatCompletionResponseEvent(
+                    event_type=ChatCompletionResponseEventType.start,
+                    delta="",
+                )
+            )
+
+            stream = await self.client.chat(
+                model=self.model,
+                messages=self._messages_to_ollama_messages(request.messages),
+                stream=True
+            )
+
+            buffer = ""
+            ipython = False
+            stop_reason = None
+
+            async for chunk in stream:
+                # check if ollama is done
+                if chunk['done']:
+                    if chunk['done_reason'] == 'stop':
+                        stop_reason = StopReason.end_of_turn
+                    elif chunk['done_reason'] == 'length':
+                        stop_reason = StopReason.out_of_tokens
+                    break
+
+                text = chunk['message']['content']
+
+                # check if its a tool call ( aka starts with <|python_tag|> )
+                if not ipython and text.startswith("<|python_tag|>"):
+                    ipython = True
+                    yield ChatCompletionResponseStreamChunk(
+                        event=ChatCompletionResponseEvent(
+                            event_type=ChatCompletionResponseEventType.progress,
+                            delta=ToolCallDelta(
+                                content="",
+                                parse_status=ToolCallParseStatus.started,
+                            ),
+                        )
+                    )
+                    buffer = buffer[len("<|python_tag|>") :]
```

Comment:
in this case `buffer` isn't even being added to?

Reply:
The assumption is that we the model always starts with `<|python_tag|>` and hence its just taking everything after that tag.

Reply:
no but `buffer` the variable isn't even being assigned to -- nothing is being to added to it?